### DISCUSSION TITLE
is it possible/how   incrementally train your code for CategoricalNB  Naive Bayes

### DISCUSSION THREAD

**Discussion by Sandy4321** [online-ml/river#774] [Q&A]
is it possible  incrementally train your code CategoricalNB  Naive Bayes 

do you have such kind of code example ?

lets say feature N3 has values : abc, hjki, deert, sdfr, qqqq, errttyy, etc

then model was trained initially 
after new values was observed : "qwer", "abc", "abc", "abc","errttyy", "qwer","qwer","qwer","qwer"
1
when and how new value "qwer" will be used for prediction 
2
how new value "qwer" willbe used for model retraining?
3
how new value "qwer" will be not used for model prediction and how prediction  is happen by  not using feature N3 when "qwer"  observed , until we have many "qwer"  observed for all classes   ?

 

when scikit  CategoricalNB   can not predict / retrain for new values in features 
https://scikit-learn.org/stable/modules/generated/sklearn.naive_bayes.CategoricalNB.html#sklearn.naive_bayes.CategoricalNB

scikit only can  deal with streaming data  for MultinomialNB, BernoulliNB, and GaussianNB
they call it by "out-of-core-naive-bayes-model-fitting"

https://scikit-learn.org/stable/modules/naive_bayes.html#out-of-core-naive-bayes-model-fitting

MultinomialNB, BernoulliNB, and GaussianNB expose a partial_fit method that can be used incrementally as done with other classifiers

my guess full streaming naive-bayes should identify new values and automatically retrain model to handle this new values .
When only when we have enough observation for particular new values , we can start to use this value for predictions
Since if certain categorical value   was observed let say 3 times for class "YES" and 50 times for class no;   it can not be used still for prediction . Until at least we have 10 times for class "YES" .???

**Comment by raphaelsty** [comment]
Hi @Sandy4321, River provides MultinomialNB, BernoulliNB, GaussianNB and ComplementNB as sklearn does but does not provide CategoricalNB. New values and features are handled by Naive Bayes (with River). Models automatically update their states to handle these values. Naive Bayes models tend to rely on the class probability rather than the class probability given the new value when there are not enough samples for a given value, you can increase the parameter `alpha` to do smoothing (i.e wait for more samples to take a decision) .

**Reply by MaxHalford** [reply]
Note that `CategoricalNB` should be implementable (is that a word?) in River. I've added a [card](https://www.notion.so/d1e86fcdf21e4deda16eedab2b3361fb?v=503f44740b8b44a99a961aa96e9e46e1&p=026ee628ee834a4cae63ab443c466bad) to the roadmap.

**Reply by Sandy4321** [reply]
@MaxHalford 
great thanks a lot 
we really need streaming data CategoricalNB
meantime , do you know what python packages to check if they have already streaming 
 data CategoricalNB  ?
it seems to be so natural implement first  streaming data CategoricalNB way before more fancier algorithms ?

**Reply by MaxHalford** [reply]
> meantime , do you know what python packages to check if they have already streaming
data CategoricalNB ?

You can use scikit-learn! [Their implementation](https://scikit-learn.org/stable/modules/generated/sklearn.naive_bayes.CategoricalNB.html) has a `partial_fit` method. I know it's not great because it doesn't adapt to new features 🤷🏼‍♂️

>  it seems to be so natural implement first streaming data CategoricalNB way before more fancier algorithms ?

Agreed, starting simple is always a good idea.

>  we really need streaming data CategoricalNB

~~I'll take a look and see if I can't get it done very quickly :)~~

Edit: I actually won't have to implement this soon if I want to do it correctly. Ideally I would like to refactor a bit of code in the `naive_bayes` module. Our current code is smart, but we've lost in readability IMO. Sometimes you only realise than when you plunge back into the code after a few months :). In the meantime, I would suggest using River's `BernoulliNB` and one-hot encoding your categories. It should do a decent job.

**Reply by Sandy4321** [reply]
1
You can use scikit-learn! Their implementation has a partial_fit method. I know it's not great because it doesn't adapt to new features 🤷🏼‍♂️

https://github.com/scikit-learn/scikit-learn/blob/0d378913b/sklearn/naive_bayes.py#L1328

the main problem with scikit-learn  CategoricalNB  partial_fit method
seems to be they do not have capability to treat new values 
2
I would suggest using River's BernoulliNB and one-hot encoding your categories. It should do a decent job.

the same problem with using one-hot / BernoulliNB 
Can RIVER treat properly new values ( not seen in one - hot) 
since one hot just will null all new values - but we need new column adding to old one hot and after proper BernoulliNB  treating this new columns
3

pls see what I wrote above about new values
`my guess full streaming naive-bayes should identify new values and automatically retrain model to handle this new values .
When only when we have enough observation for particular new values , we can start to use this value for predictions
Since if certain categorical value was observed let say 3 times for class "YES" and 50 times for class no; it can not be used still for prediction . Until at least we have 10 times for class "YES" .???`

**Reply by Sandy4321** [reply]
So , we need to add new parameter : minimum number of observation for new value per class  to be used for prediction ??

**Reply by Sandy4321** [reply]
another problem with scikit - they do not add new labels
#https://scikit-learn.org/stable/modules/generated/sklearn.naive_bayes.CategoricalNB.html#sklearn.naive_bayes.CategoricalNB.partial_fit
import numpy as np
rng = np.random.RandomState(1)
X1 = rng.randint(5, size=(6, 100))
y1 = np.array([1, 2, 3, 4, 5, 6])
from sklearn.naive_bayes import CategoricalNB
clf = CategoricalNB()
clf.fit(X1, y1)
X2 = rng.randint(5, size=(6, 100))
y2 = np.array([6, 5, 4, 3, 2, 1])
clf.partial_fit(X2, y2)

X3 = rng.randint(7, size=(6, 100))
y3 = np.array([12, 21, 3, 4, 5, 6])
clf.partial_fit(X3, y3)

print('predict X1')
print(clf.predict(X1))
print('predict X2')
print(clf.predict(X2))
print('predict X3')
print(clf.predict(X3))

predict X1
[1 2 3 4 5 6]
predict X2
[6 5 4 3 2 1]
predict X3
[2 3 3 4 5 6]

new labels 12 and 21 were detected as old labels 2 and 3

**Reply by Sandy4321** [reply]
I also have doubts how they treat new values from line 
X3 = rng.randint(7, size=(6, 100))

old data was from 0 to 5
but new data is from 0 to 7

**Reply by Sandy4321** [reply]
z = np.array( [ [0,1,2] , [ 1,2,2] , [ 0,1,2 ], [0,2,2] , [1,1,1]  , [0,0,1] ] )
y = [ 0,0,0,0,1,1]

clf = CategoricalNB()
clf.fit(z, y)
print('befor adding new data classes')
print(clf.classes_)
print('before adding new data feature 1')
print(clf.feature_log_prob_[0])
print('befor adding new data  feature 2')
print(clf.feature_log_prob_[1])
print('predict z' )
print(clf.predict(z))

v  = np.array( [ [11,2,1] , [ 11,1,2] , [ 11,2,1 ], [11,2,0] , [11,1,1] , [11, 2, 1] ] )
b = [ 1,1,1,1,1,45]

clf.partial_fit(v, b)
print('predict v' )
print(clf.predict(v))

print('after adding new data classes')
print(clf.classes_)

print('after adding new data feature 1')
print(clf.feature_log_prob_[0])
print('after adding new data  feature 2')
print(clf.feature_log_prob_[1])

**Reply by Sandy4321** [reply]
output is

`befor adding new data classes
[0 1]
before adding new data feature 1
[[-0.40546511 -1.09861229]
 [-0.69314718 -0.69314718]]
befor adding new data  feature 2
[[-1.94591015 -0.84729786 -0.84729786]
 [-0.91629073 -0.91629073 -1.60943791]]
predict z
[0 0 0 0 1 1]
predict v
[1 1 1 1 1 1]
after adding new data classes
[0 1]
after adding new data feature 1
[[-1.44691898 -2.14006616 -2.83321334 -2.83321334 -2.83321334 -2.83321334
  -2.83321334 -2.83321334 -2.83321334 -2.83321334 -2.83321334 -2.14006616]
 [-2.2512918  -2.2512918  -2.94443898 -2.94443898 -2.94443898 -2.94443898
  -2.94443898 -2.94443898 -2.94443898 -2.94443898 -2.94443898 -1.15267951]]
after adding new data  feature 2
[[-2.07944154 -0.98082925 -0.69314718]
 [-1.60943791 -0.91629073 -0.91629073]]`

**Reply by Sandy4321** [reply]
output is 
befor adding new data classes
[0 1]
before adding new data feature 1
[[-0.40546511 -1.09861229]
 [-0.69314718 -0.69314718]]
befor adding new data  feature 2
[[-1.94591015 -0.84729786 -0.84729786]
 [-0.91629073 -0.91629073 -1.60943791]]
predict z
[0 0 0 0 1 1]
predict v
[1 1 1 1 1 1]
after adding new data classes
[0 1]
after adding new data feature 1
[[-1.44691898 -2.14006616 -2.83321334 -2.83321334 -2.83321334 -2.83321334
  -2.83321334 -2.83321334 -2.83321334 -2.83321334 -2.83321334 -2.14006616]
 [-2.2512918  -2.2512918  -2.94443898 -2.94443898 -2.94443898 -2.94443898
  -2.94443898 -2.94443898 -2.94443898 -2.94443898 -2.94443898 -1.15267951]]
after adding new data  feature 2
[[-2.07944154 -0.98082925 -0.69314718]
 [-1.60943791 -0.91629073 -0.91629073]]

**Reply by Sandy4321** [reply]
seems to be they not ignore new values, but add it by very strange method - add many dummy values

**Reply by MaxHalford** [reply]
Please @Sandy4321, can you put some effort into your messages? At least highlight and format the code properly. Currently it's very painful for me to understand what you're trying to communicate.

**Reply by Sandy4321** [reply]
though problem with new classes is here 
befor adding new data classes
[0 1]
after adding new data classes
[0 1]
classes are the same before and after adding new class

**Reply by Sandy4321** [reply]
though it is not clear what they mean under
min_categories

https://scikit-learn.org/0.24/whats_new/v0.24.html?highlight=categoricalnb

Enhancement Adds a parameter min_categories to naive_bayes.CategoricalNB that allows a minimum number of categories per feature to be specified. This allows categories unseen during training to be accounted for. #16326 by George Armstrong.

**Comment by Sandy4321** [comment]
see also my comments to 
https://github.com/scikit-learn/scikit-learn/pull/16326

about new data error-
import numpy as np
rng = np.random.RandomState(1)
X = rng.randint(5, size=(4, 7))
y = np.array([1, 0, 1, 1])
from sklearn.naive_bayes import CategoricalNB
clf = CategoricalNB()
clf.min_categories = 2
clf.fit(X, y)
X= 
array([[3, 4, 0, 1, 3, 0, 0],
       [1, 4, 4, 1, 2, 4, 2],
       [4, 3, 4, 2, 4, 2, 4],
       [1, 1, 0, 1, 1, 1, 1]])
test = [[4, 3, 4, 2, 4, 2, 123]]
print(clf.predict(test))

Traceback (most recent call last):
  Debug Console, prompt 30, line 11
    clf.fit(X1, y1)
  File "e:\Python environemnts\MYENVS\may9_sklearn0242\Lib\site-packages\sklearn\naive_bayes.py", line 75, in predict
    jll = self._joint_log_likelihood(X)
  File "e:\Python environemnts\MYENVS\may9_sklearn0242\Lib\site-packages\sklearn\naive_bayes.py", line 1303, in _joint_log_likelihood
    jll += self.feature_log_prob_[i][:, indices].T
builtins.IndexError: index 123 is out of bounds for axis 1 with size 5

**Reply by Sandy4321** [reply]
also it limiting that scikit  represent values only in integer form

when actual data as rule is in string form
 then as they suggests to use OrdinalEncoder  
that it become cumbersome to update  OrdinalEncoder   each time new data values coming ...
better to do native string input data with proper treating new values : retrain and and when we have enough new values predict with using new values
but until it happen - just ignore features with new values for given row prediction ?   

https://github.com/gwarmstrong
@gwarmstrong
@prabhat00155
@madhuracj
https://github.com/madhuracj
https://github.com/prabhat00155
https://github.com/glemaitre

@glemaitre
https://github.com/jnothman
@jnothman
Though a more efficient solution would be to use an OrdinalEncoder on your data, which will correctly identify the actual number of categories in the data, and provide the ability to set values not seen in training data by using the unknown_value in OrdinalEncoder. Then min_categories can be set to number_of_features_in_training_data + 1

**Reply by Sandy4321** [reply]
maybe lets do it one by one

**Reply by Sandy4321** [reply]
you wrote:
------------------------------------------------------------------
In the meantime, I would suggest using River's
BernoulliNB
and one-hot encoding yourcategories. It should do a decent job.
-------------------------------------------------------------------

but to do on line onehot encoder is more difficult than proper categorical NB?
in any case may you share code example for such at solution , pls
When new values treated properly by one hot encoder - 
new columns are added to one hot data as new values coming ?

**Reply by Sandy4321** [reply]
seems to be you can use code ideas from https://github.com/geofizx/Streaming-Naive-Bayes

**Comment by ChandraPrakash-Bathula** [comment]
CategoricalNB in scikit-learn does not have a direct method for incremental training or partial fitting like MultinomialNB, BernoulliNB, and GaussianNB. I think, CategoricalNB does not provide a partial_fit method.

However, you can achieve incremental training for categorical data by manually updating the parameters of the model when new data becomes available.
